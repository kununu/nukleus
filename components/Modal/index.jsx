import React from 'react';
import PropTypes from 'prop-types';
import AriaModal from 'react-aria-modal';

import styles from './index.scss';

import Button from '../Button';

export default class Modal extends React.Component {

  static propTypes = {
    actionText: PropTypes.string,
    cancelText: PropTypes.string,
    children: PropTypes.element.isRequired,
    closeText: PropTypes.string,
    onAction: PropTypes.func,
    onExit: PropTypes.func.isRequired,
    open: PropTypes.bool.isRequired,
    titleText: PropTypes.string.isRequired
  }

  static defaultProps = {
    actionText: '',
    cancelText: '',
    closeText: 'Close',
    onAction: () => {},
    verticallyCenter: true
  };

  onAction = ev => {
    this.props.onAction(ev);
  }

  onExit = ev => {
    this.props.onExit(ev);
  }

  renderFooter () {
    const {
      actionText,
      cancelText
    } = this.props;

    if (actionText || cancelText) {
      return (
        <footer className={styles.modalFooter}>
          {actionText && <Button
            type="primary"
            text={actionText}
            onClick={this.props.onAction} />}
          {cancelText && <Button
            type="secondary"
            text={cancelText}
            onClick={this.props.onExit} />}
        </footer>
      );
    }

    return null;
  }

  render () {
    return (
      this.props.open ? <AriaModal
        {...this.props}>
        <section className={styles.modal}>
          <header className={styles.modalHeader}>
            <span className={styles.modalTitle}>
              {this.props.titleText}
            </span>
            <button
              type="button"
              className={styles.closeButton}
              onClick={this.props.onExit}>
              <span role="presentation">
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  className={styles.closeIcon}
                  xmlnsXlink="http://www.w3.org/1999/xlink"
                  xmlSpace="preserve"
                  version="1.1"
                  width="100%"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 16 16">
                  <title>{this.props.closeText}</title>
                  <path d="M9,7 L9,3.55271368e-15 L7,3.55271368e-15 L7,7 L7.10542736e-15,7 L7.10542736e-15,9 L7,9 L7,16 L9,16 L9,9 L16,9 L16,7 L9,7 Z" transform="translate(8.000000, 8.000000) rotate(-45.000000) translate(-8.000000, -8.000000) " />
                </svg>
              </span>
            </button>
          </header>
          <div className={styles.modalBody}>{this.props.children}</div>
          {this.renderFooter()}
        </section>
      </AriaModal> : null
    );
  }
}
