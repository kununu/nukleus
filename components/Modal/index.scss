@import '~kununu-theme/scss/base/variables';

.modal {
  background: $white;
  z-index: 10;
  box-shadow: 0 0 25px rgba(0, 0, 0, .3);
  min-width: 400px;
  opacity: .01;
  transform: scale(.85);
  transition: transform .15s ease-in-out, opacity .15s ease-in-out;
  transition-delay: .05s;
}

.modalHasEntered {
  opacity: 1;
  transform: scale(1);
}

.underlay {
  opacity: .01;
  transition: opacity .1s ease-in-out;
}

.underlayHasEntered {
  opacity: 1;
}

.modalHeader {
  width: 100%;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 15px 10px;
  border-bottom: .1px solid $gray-base-20;
}

.modalFooter {
  text-align: right;
  padding: 15px 10px;
  display: flex;
  justify-content: flex-end;
  margin: 0 -5px;

  > * {
    margin: 0 5px;
    min-width: 80px;
  }
}

.modalTitle {
  font-size: 16px;
  line-height: 21px;
  font-weight: normal;
}

.modalBody {
  padding: 10px;
}

.closeButton {
  width: 16px;
  height: 16px;
  border: 0;
  background: transparent;
  padding: 0;

  &:focus {
    outline: 0;
  }

  &:active {
    svg {
      fill: $blue-lighter;
    }
  }
}

.closeIcon {
  display: block;
  width: 16px;
  height: 16px;
  stroke: none;
  fill: $blue;
}
