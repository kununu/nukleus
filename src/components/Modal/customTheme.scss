@import '~@kununu/kununu-theme-v2/scss/base/variables';

.modal {
  background: $white;
  z-index: 10;
  box-shadow: 0 0 25px rgba(0, 0, 0, .3);
  min-width: 320px;
  opacity: .01;
  transform: scale(.85);
  transition: transform .15s ease-in-out, opacity .15s ease-in-out;
  transition-delay: .05s;
  border-radius: 10px;

  @media (min-width: 480px) {
    min-width: 400px;
  }
}

.modalIsOpen {
  opacity: 1;
  transform: scale(1);
}

.modalHeader {
  box-sizing: border-box;
  width: 100%;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 20px;
}

.modalFooter {
  text-align: right;
  padding: 20px;
  display: flex;
  justify-content: flex-end;
  margin: 0 -5px;
}

.modalBody {
  padding: 0 20px;
}

.modalTitle {
  margin: 0;
  font-size: 18px;
}

.modalCloseButton {
  width: 12px;
  height: 12px;
  border: 0;
  background: transparent;
  padding: 0;

  &:focus {
    outline: 0;
  }
}

.modalCloseIcon {
  display: block;
  width: 12px;
  height: 12px;
  stroke: none;
  fill: black;
}
