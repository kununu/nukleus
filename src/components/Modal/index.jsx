import React from 'react';
import PropTypes from 'prop-types';
import AriaModal from 'react-aria-modal';

import ThemeContext from 'utils/themeContext';
import themeable from 'utils/theming';

import Button from '../Button';

import styles from './index.module.scss';


export default class Modal extends React.Component {
  static propTypes = {
    actionText: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element,
    ]),
    cancelText: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element,
    ]),
    children: PropTypes.element.isRequired,
    closeText: PropTypes.string,
    onAction: PropTypes.func, // This should be a promise, so that onExit can be executed on success (nicer animation)
    onExit: PropTypes.func.isRequired,
    open: PropTypes.bool.isRequired,
    titleText: PropTypes.string.isRequired,
  }

  static defaultProps = {
    actionText: '',
    cancelText: '',
    closeText: 'Close',
    onAction: () => {},
  };

  constructor (props) {
    super(props);

    this.state = {isOpen: false};

    // needed to have real binding to get tests working
    this.onExit = this.onExit.bind(this);
    this.onAction = this.onAction.bind(this);
    this.onEnter = this.onEnter.bind(this);
  }

  onAction (ev) {
    const {onAction} = this.props;

    onAction(ev)
      .then(() => this.onExit());
  }

  onExit (ev) {
    // First set state for nicer animation
    this.setState({isOpen: false}, () => {
      setTimeout(() => {
        // Call parent onExit
        const {onExit} = this.props;

        onExit(ev);
      }, 250); // Time until animation has finished
    });
  }

  onEnter () {
    this.setState({isOpen: true});
  }

  renderFooter (theme) {
    const {
      actionText,
      cancelText,
    } = this.props;

    if (actionText || cancelText) {
      return (
        <footer className={theme('modalFooter')}>
          {actionText && (
          <Button
            htmlType="button"
            onClick={this.onAction}
            text={actionText}
            type="primary"
            id="nukleus-modal-action-btn"
          />
          )}
          {cancelText && (
          <Button
            htmlType="button"
            onClick={this.onExit}
            text={cancelText}
            type="secondary"
            id="nukleus-modal-cancel-btn"
          />
          )}
        </footer>
      );
    }

    return null;
  }

  render () {
    const {
      isOpen,
    } = this.state;
    const {
      children,
      closeText,
      open,
      titleText,
    } = this.props;

    const titleId = 'nukleus-modal-title'; // will be used for aria-labelledby
    const underlayClass = isOpen ? 'underlayHasEntered' : '';
    const isOpenClass = isOpen ? 'modalIsOpen' : '';
    const overrideProps = {
      ...this.props,
      titleId,
      verticallyCenter: true, // here we override react-aria-modal defaults
    };

    return (
      <ThemeContext.Consumer>
        {(context) => {
          const theme = themeable({...styles, ...context});

          return (
            open ? (
              <AriaModal
                {...overrideProps}
                onExit={this.onExit}
                underlayClass={theme('underlay', underlayClass)}
                onEnter={this.onEnter}
              >
                <section className={theme('modal', isOpenClass)}>
                  <header className={theme('modalHeader')}>
                    <h1
                      id={titleId}
                      className={theme('modalTitle')}
                    >
                      {titleText}
                    </h1>
                    <button
                      className={theme('modalCloseButton')}
                      id="nukleus-modal-close"
                      onClick={this.onExit}
                      title={closeText}
                      type="button"
                    >
                      <span role="presentation">
                        <svg
                          xmlns="http://www.w3.org/2000/svg"
                          className={theme('modalCloseIcon')}
                          xmlnsXlink="http://www.w3.org/1999/xlink"
                          xmlSpace="preserve"
                          version="1.1"
                          width="100%"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 16 16"
                        >
                          <path
                            d="M9,7 L9,3.55271368e-15 L7,3.55271368e-15 L7,7 L7.10542736e-15,7 L7.10542736e-15,9 L7,9 L7,16 L9,16 L9,9 L16,9 L16,7 L9,7 Z"
                            transform="translate(8.000000, 8.000000) rotate(-45.000000) translate(-8.000000, -8.000000) "
                          />
                        </svg>
                      </span>
                    </button>
                  </header>
                  <div className={theme('modalBody')}>{children}</div>
                  {this.renderFooter(theme)}
                </section>
              </AriaModal>
            ) : null
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
