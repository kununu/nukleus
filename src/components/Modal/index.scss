@import '~@kununu/kununu-theme-v2/scss/base/variables';

.modal {
  background: $white;
  z-index: 10;
  box-shadow: 0 0 25px rgba(0, 0, 0, .3);
  min-width: 320px;
  opacity: .01;
  transform: scale(.85);
  transition: transform .15s ease-in-out, opacity .15s ease-in-out;
  transition-delay: .05s;

  @media (min-width: 480px) {
    min-width: 400px;
  }
}

.modalIsOpen {
  opacity: 1;
  transform: scale(1);
}

.underlay {
  opacity: .01;
  transition: opacity .1s ease-in-out;
}

.underlayHasEntered {
  opacity: 1;
}

.modalHeader {
  width: 100%;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 25px 20px;
  border-bottom: .1px solid $gray-base-20;
}

.modalFooter {
  text-align: right;
  padding: 25px 20px;
  display: flex;
  justify-content: flex-end;
  margin: 0 -5px;

  > * {
    margin: 0 5px;
    min-width: 80px;
  }
}

.modalTitle {
  margin: 0;
}

.modalBody {
  padding: 20px;
}

.modalCloseButton {
  width: 16px;
  height: 16px;
  border: 0;
  background: transparent;
  padding: 0;

  &:focus {
    outline: 0;
  }

  &:active {
    svg {
      fill: $blue-lighter;
    }
  }
}

.modalCloseIcon {
  display: block;
  width: 16px;
  height: 16px;
  stroke: none;
  fill: $blue;
}
