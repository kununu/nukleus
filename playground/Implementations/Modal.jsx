import React from 'react';

import Modal from 'components/Modal';
import Button from 'components/Button';

export default class ModalImplementation extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      modalOpen: false,
      success: false
    };
  }

  render () {
    return (
      <div>
        <h1>Modal is open: {JSON.stringify(this.state.modalOpen)}</h1>
        <Button
          type="primary"
          text="Open Modal"
          onClick={() => this.setState({modalOpen: true})} />
        <Modal
          titleText="dafuq"
          open={this.state.modalOpen}
          state={this.state}
          actionText="Ok"
          cancelText="Cancel"
          onCancel={() => this.setState({modalOpen: false})}
          onAction={() => this.setState({
            modalOpen: false,
            success: true
          })}>
          <p>O rly ?</p>
        </Modal>
      </div>
    );
  }

}
