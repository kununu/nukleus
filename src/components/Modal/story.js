import React, {Component} from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {withInfo} from '@storybook/addon-info';
import {withKnobs, text} from '@storybook/addon-knobs/react';

import Button from '../Button';

import Modal from './index';

class ModalExample extends Component {
  constructor (props) {
    super(props);

    this.state = {modalOpen: false}; // eslint-disable-line react/no-unused-state
  }

  render () {
    const {modalOpen} = this.state;

    return (
      <div>
        <p>Use this button, to show the modal</p>
        <Button
          type="primary"
          text="Open Modal"
          onClick={() => this.setState({modalOpen: true})}
        />
        <Modal
          titleText={text('titleText', 'Nukleus Modal')}
          open={modalOpen}
          state={this.state}
          actionText={text('actionText', 'Ok')}
          cancelText={text('cancelText', 'Cancel')}
          onExit={() => this.setState({modalOpen: false})}
          onAction={() => this.setState({
            modalOpen: false,
            success: true, // eslint-disable-line react/no-unused-state
          })}
        >
          <p>{text('dynamic text content', 'dynamic text content')}</p>
        </Modal>
      </div>
    );
  }
}

storiesOf('Modal', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('The modal component! You can wrap any children components that you like in this fabulous modal.')(story)(context))
  .add('basics', () => (
    <ModalExample />
  ));
