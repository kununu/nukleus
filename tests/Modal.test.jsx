import React from 'react';
import toJson from 'enzyme-to-json';
import {mount} from 'enzyme';
import Modal from 'Modal'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

test('Renders Modal without crashing', () => {
  const component = mount(
    <Modal
      actionText="Ok"
      cancelText="Cancel"
      closeText="Close"
      applicationNode="#test"
      onAction={() => {}}
      onExit={() => {}}
      open
      titleText="Modal">
      <p>Hello world</p>
    </Modal>
  );

  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('Calls onEnter, when the modal is mounted', () => {
  const onEnter = window.spyOn(Modal.prototype, 'onEnter');
  mount(
    <Modal
      actionText="Ok"
      cancelText="Cancel"
      closeText="Close"
      applicationNode="#test"
      onAction={() => {}}
      onExit={() => {}}
      open
      titleText="Modal">
      <p>Hello world</p>
    </Modal>
  );

  expect(onEnter).toHaveBeenCalled();
});

test('Closes the modal, when the close button is clicked', () => {
  const onExit = window.spyOn(Modal.prototype, 'onExit');
  const component = mount(
    <Modal
      actionText="Ok"
      cancelText="Cancel"
      closeText="Close"
      applicationNode="#test"
      onAction={() => {}}
      onExit={() => {}}
      open
      titleText="Modal">
      <p>Hello world</p>
    </Modal>
  );

  component.find('#nukleus-modal-close').simulate('click');
  expect(onExit).toHaveBeenCalled();
});

/**
 * It would be cool, if we could rewrite this test in the future, so that it also tests,
 * if onExit was called, when onAction is a promise
 */
test('Fires the onAction when the action button is clicked', done => {
  const onAction = window.spyOn(Modal.prototype, 'onAction');

  const component = mount(
    <Modal
      actionText="Ok"
      cancelText="Cancel"
      closeText="Close"
      applicationNode="#test"
      onAction={() => {}}
      onExit={() => {}}
      open
      titleText="Modal">
      <p>Hello world</p>
    </Modal>
  );

  component.find('footer button').first().simulate('click');

  setTimeout(() => {
    expect(onAction).toHaveBeenCalled();
    done();
  }, 250);
});

test('Fires the onExit when the close button is clicked', done => {
  const onExit = window.spyOn(Modal.prototype, 'onExit');

  const component = mount(
    <Modal
      actionText="Ok"
      cancelText="Cancel"
      closeText="Close"
      applicationNode="#test"
      onAction={() => {}}
      onExit={() => {}}
      open
      titleText="Modal">
      <p>Hello world</p>
    </Modal>
  );

  component.find('footer button').last().simulate('click');

  setTimeout(() => {
    expect(onExit).toHaveBeenCalled();
    done();
  }, 250);
});
