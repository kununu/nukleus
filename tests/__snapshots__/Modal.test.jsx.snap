exports[`test Renders Modal without crashing 1`] = `
<Modal
  actionText="Ok"
  applicationNode="#test"
  cancelText="Cancel"
  closeText="Close"
  onAction={[Function]}
  onExit={[Function]}
  open={true}
  titleText="Modal">
  <Modal
    actionText="Ok"
    applicationNode="#test"
    cancelText="Cancel"
    closeText="Close"
    dialogId="react-aria-modal-dialog"
    escapeExits={true}
    focusTrapPaused={false}
    includeDefaultStyles={true}
    onAction={[Function]}
    onEnter={[Function]}
    onExit={[Function]}
    open={true}
    scrollDisabled={true}
    titleId="nukleus-modal-title"
    titleText="Modal"
    underlayClass="underlay underlayHasEntered"
    underlayClickExits={true}
    underlayColor="rgba(0,0,0,0.5)"
    underlayProps={Object {}}
    verticallyCenter={true}>
    <FocusTrap
      _createFocusTrap={[Function]}
      active={true}
      focusTrapOptions={
        Object {
          "initialFocus": undefined,
        }
      }
      paused={false}
      tag="div">
      <div>
        <div
          className="underlay underlayHasEntered"
          onClick={[Function]}
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "background": "rgba(0,0,0,0.5)",
              "cursor": "pointer",
              "height": "100%",
              "left": 0,
              "overflowX": "hidden",
              "overflowY": "auto",
              "position": "fixed",
              "textAlign": "center",
              "top": 0,
              "width": "100%",
              "zIndex": 1050,
            }
          }>
          <div
            style={
              Object {
                "display": "inline-block",
                "height": "100%",
                "verticalAlign": "middle",
              }
            } />
          <div
            aria-labelledby="nukleus-modal-title"
            id="react-aria-modal-dialog"
            role="dialog"
            style={
              Object {
                "cursor": "default",
                "display": "inline-block",
                "maxWidth": "100%",
                "outline": null,
                "textAlign": "left",
                "top": 0,
                "verticalAlign": "middle",
              }
            }>
            <section
              className="modal isOpen">
              <header
                className="modalHeader">
                <h1
                  className="modalTitle"
                  id="nukleus-modal-title">
                  Modal
                </h1>
                <button
                  className="closeButton"
                  id="nukleus-modal-close"
                  onClick={[Function]}
                  title="Close"
                  type="button">
                  <span
                    role="presentation">
                    <svg
                      className="closeIcon"
                      preserveAspectRatio="xMidYMid meet"
                      version="1.1"
                      viewBox="0 0 16 16"
                      width="100%"
                      xmlSpace="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      xmlnsXlink="http://www.w3.org/1999/xlink">
                      <path
                        d="M9,7 L9,3.55271368e-15 L7,3.55271368e-15 L7,7 L7.10542736e-15,7 L7.10542736e-15,9 L7,9 L7,16 L9,16 L9,9 L16,9 L16,7 L9,7 Z"
                        transform="translate(8.000000, 8.000000) rotate(-45.000000) translate(-8.000000, -8.000000) " />
                    </svg>
                  </span>
                </button>
              </header>
              <div
                className="modalBody">
                <p>
                  Hello world
                </p>
              </div>
              <footer
                className="modalFooter">
                <Button
                  customTheme=""
                  disabled={false}
                  fullWidth={false}
                  htmlType="button"
                  link={null}
                  mobileFullWidth={false}
                  onClick={[Function]}
                  outline={false}
                  text="Ok"
                  type="primary">
                  <button
                    className="button primary"
                    disabled={false}
                    onClick={[Function]}
                    type="button">
                    Ok
                  </button>
                </Button>
                <Button
                  customTheme=""
                  disabled={false}
                  fullWidth={false}
                  htmlType="button"
                  link={null}
                  mobileFullWidth={false}
                  onClick={[Function]}
                  outline={false}
                  text="Cancel"
                  type="secondary">
                  <button
                    className="button secondary"
                    disabled={false}
                    onClick={[Function]}
                    type="button">
                    Cancel
                  </button>
                </Button>
              </footer>
            </section>
          </div>
        </div>
      </div>
    </FocusTrap>
  </Modal>
</Modal>
`;
